Workaround shader compiler bugs with degenerate switch statements

Imberflur opened the pull request:
**Connections**
Fixes https://github.com/gfx-rs/wgpu/issues/4514
Fixes https://github.com/gfx-rs/wgpu/issues/4485

**Description**

As discussed in https://github.com/gfx-rs/wgpu/issues/4485, FXC has a bug when handling switches with all cases leading to one body (it deletes the switch). It was also mentioned that some glsl consumers might not handle this scenario well either (but I don't know which ones that might be).

So we lower these to `do {} while(false);` loops in hlsl-out and glsl-out.

While FXC already can't handle continue statements in switches (see #4485), I ended up including logic to properly forward continue statements to the outer loop here, since there would otherwise be regressions in glsl-out and maybe hlsl-out with DXC.

An alternative solution could be to switch on a known constant and have a dummy case with an empty body that is never hit. This is a bit less cautious in terms of avoiding potential compiler bugs since we are still switching on a constant. But the implementation would be simpler and the only issues I have observed so far with FXC are when all cases lead to a single body.

**Testing**

Some of these scenarios are already covered by existing naga snapshot tests, I extended `naga/tests/in/control-flow.wgsl` with a few more cases.

I also included some regression tests for the linked issues that will execute the shaders and assert on their output. Additionally, a new mechanism is introduced into the test framework for forcing the usage of FXC when on the Dx12 backend.

I ran `cargo xtask test` on a windows machine and a separate linux machine.

<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

Imberflur committed:
[naga hlsl-out glsl-out] Work around backend loop/switch bugs.

Introduce a new module, `naga::back::continue_forward`, containing
shared code for rendering Naga `Continue` statements as backend
`break` statements and assignments to introduced `bool` locals.
See the module's documentation for details.

- [hlsl-out] Transform degenerate single body switches into `do-while`
  loops. Properly render `Continue` statements enclosed by
  `Switch` statements enclosed by `Loop` statements.

- [glsl-out] Transform degenerate single body switches into `do-while`
  loops.

Improve `naga xtask validate spv` error message.

Fixes #4485.
Fixes #4514.

Imberflur commented:
```
stderr:
error: line 127: Case construct that targets '74[%74]' has invalid branch to block '55[%55]' (not another case construct, corresponding merge, outer loop merge or outer loop continue)
  %74 = OpLabel
```

I guess this is one of the nested test cases I added.

Imberflur commented:
> ```
> stderr:
> error: line 127: Case construct that targets '74[%74]' has invalid branch to block '55[%55]' (not another case construct, corresponding merge, outer loop merge or outer loop continue)
>   %74 = OpLabel
> ```
> 
> I guess this is one of the nested test cases I added.

Found minimal case https://github.com/gfx-rs/wgpu/issues/5658 and was able to avoid the issue here without affecting the aspects I was trying to test.

Imberflur commented on naga/src/back/glsl/mod.rs:
```diff
@@ -2077,42 +2083,110 @@ impl<'a, W: Write> Writer<'a, W> {
-                        crate::SwitchValue::U32(value) => write!(self.out, "{l2}case {value}u:")?,
-                        crate::SwitchValue::Default => write!(self.out, "{l2}default:")?,
-                    }
+                // Some glsl consumers may not handle switches with a single body correctly.
+                // So write this as a `do {} while(false);` loop instead.
+                // This is the same fix as for FXC in hlsl, although here it
+                // wasn't specifically tested against any problematic consumers.
+                // See https://github.com/gfx-rs/wgpu/issues/4514
+                let one_body = cases
+                    .iter()
+                    .rev()
+                    .skip(1)
+                    .all(|case| case.fall_through && case.body.is_empty());
```
One issue I just realized is what if none of these switch cases would have matched, I need to check if that's possible.

Imberflur replied:
Nvm looks like naga's switch statement requires a default case!

Imberflur replied:
Maybe I should note that here?

Imberflur replied:
I added a note

cwfitzgerald commented on tests/src/init.rs:
```diff
@@ -27,7 +27,12 @@ pub fn initialize_instance() -> Instance {
         Backends::all()
     };
-    let dx12_shader_compiler = wgpu::util::dx12_shader_compiler_from_env().unwrap_or_default();
+    // TODO: should env override force_fxc?
```
No - tests need to be able to force demote it.

Imberflur replied:
Thanks! Removed TODO comment and added a small note about this.

cwfitzgerald reviewed (approved):
Wgpu side looks amazing, minus a nit

Imberflur commented:
Let me know if/when conflicts need to be fixed.

jimblandy commented on naga/src/back/continue_forward.rs:
```diff
@@ -0,0 +1,159 @@
+    ///
+    /// `continue` should apply to the current loop.
+    ///
+    /// * When entering a nested switch, add a `Forward` state to the stack.
```
What is `Forward`? Docs are out of date?

Imberflur replied:
Yeah, looks out of date. This should be `Nesting::Switch` (or just `Switch`) (confirmed by looking at the behavior in `ContinueCtx::enter_switch`).

jimblandy commented:
@Imberflur Started reviewing this. My impression so far is that this is excellent work. I especially appreciate the execution tests and the docs.

jimblandy commented:
@Imberflur Could you check out the docs I just pushed in ~~4c10a41~~ [b185798](https://github.com/gfx-rs/wgpu/pull/5654/commits/b1857983649f9e219b7e09f91f98f79f5e9af99f)?

jimblandy commented:
> Let me know if/when conflicts need to be fixed.

It seems like this needs a rebase.

edit: I did the rebase.

Imberflur commented on naga/src/back/continue_forward.rs:
```diff
@@ -0,0 +1,204 @@
+//! - We generate code for [`Continue`] statements that sets a `bool` local to
+//!   `true` and does a `break`.
+//!
+//! - When a [`Switch`] statement contains such a [`Continue`] statement:
+//!
+//!   - If it's the outermost such [`Switch`] within the targeted [`Loop`], we
+//!     declare the `bool` local, initialized to `false`, and immediately after
+//!     the `switch`, we check the local and do a `continue` if it's set.
+//!
+//!   - If the [`Switch`] is nested within other [`Switch`]es, then after the
+//!     generated `switch`, we check the local (which we know was declared
+//!     before the surrounding `switch`) and do a `break` if it's set.
```
As a nit, I would say that this line might be clarified:

> When a [`Switch`] statement contains such a [`Continue`] statement:

To me, this could imply that the following points are only applied for switches containing continue statements, when it is really applied for all switches (that are inside a loop and can't use `continue` directly) since we don't do anything to detect whether a switch statement contains `continue`s.

But FWIW I think the false implication is fairly weak and not significant to understanding the mechanisms here and I'm happy with the current text.

jimblandy replied:
It's a fair point - it might make people expect we do such a scan somewhere. I made the docs describe what we really do.

Imberflur reviewed (commented):
The new docs look great! Thanks for adding them.

I included one note for consideration.

P.S. The "forwarding" terminology is somewhat ad hoc, so feel free to suggest something else (like "deferring") if you have opinions there.

Imberflur commented:
I'm somewhat curious about what conflicts came up in the rebase (in the interest of double checking them). But I can always re-run the rebase locally to find them.

jimblandy commented:
> I'm somewhat curious about what conflicts came up in the rebase (in the interest of double checking them). But I can always re-run the rebase locally to find them.

Only the absence of a `cache` field in the pipeline descriptors.

jimblandy reviewed (changes requested):
This looks great, and I think it's correct, but I'd like to see it simplified in two ways:

- Don't bother counting nesting depth; just always push/pop new entries on the stack. It took me a bit to realize that the depth counters were just an optimization, not something that actually affected the code generated.

- Eliminate `ContinueCtx::next_id`, let `enter_switch` take a `&mut Namer` argument, and just use the namer to generate unique names directly. It took me a while to be sure that the names you were generating couldn't conflict; if the code just uses the namer, then that question won't even arise.

  This means that `Nesting::Switch` will need to actually own the `String`. I think it will need to be an `Rc<String>`, both so that nested `Switch` variants can share it, and also so that `exit_switch` can return it regardless of whether it's been popped. `ExitControl` flow will need to hold a clone of the `String`.

jimblandy commented:
I played with having the HLSL and GLSL writers just save the current nesting state in a local variable, and restore it on exit, so you could replace `ContinueCtx::stack` with a single `Nesting` value. But the API doesn't end up as nice as the one you've got, so I thought it would end up being more distracting in the end.

jimblandy commented:
@Imberflur The only other suggestion I might have - and maybe this is just too hairy to be worth it - is that we might put a flag in `Nesting::Switch` to say whether it actually encountered any `Continue` statements that needed to be forwarded. If not, then there's no need to generate the check after the `switch`. Omitting those checks might make it easier for dumb drivers to see that the bool is truly unused.

Imberflur commented:
>  just use the namer to generate unique names directly

Oh that sounds great! I didn't realize there was a mechanism for this.

Imberflur commented:
@jimblandy  I implemented suggestions:

* Removed nesting depth counter.
* Used `Namer` and `Rc<String>` for variable names.
* Omitted check generation if no continue statements are found.
* Rebased

jimblandy reviewed (approved):
Okay, I fiddled with the docs some more, and cleaned up a bit of the control flow, and I think this ready to land.

Imberflur commented:
Thanks!
